dvc/repo/experiments/executor/local.py

```diff
@@ -152,18 +167,26 @@ def __init__(self, *args, **kwargs):
     def from_stash_entry(
         cls,
         repo: "Repo",
-        stash_rev: str,
         entry: "ExpStashEntry",
         **kwargs,
     ):
         root_dir = repo.scm.root_dir
-        executor = cls._from_stash_entry(
-            repo, stash_rev, entry, root_dir, **kwargs
+        executor: "WorkspaceExecutor" = cls._from_stash_entry(
+            repo, entry, root_dir, **kwargs
         )
         logger.debug("Init workspace executor in '%s'", root_dir)
         return executor
 
-    def init_git(self, scm: "Git", branch: Optional[str] = None):
+    def init_git(
+        self,
+        scm: "Git",
+        stash_rev: str,
+        entry: "ExpStashEntry",
+        branch: Optional[str] = None,
+    ):
+        scm.set_ref(EXEC_HEAD, entry.head_rev)
+        scm.set_ref(EXEC_MERGE, stash_rev)
+        scm.set_ref(EXEC_BASELINE, entry.baseline_rev)
```

Comment:
@karajan1001 this is still a problem where we have a lot of duplicated code (in all of the different executor init methods). If you want to keep this in the executor classes we can do that, but the common code should be refactored so that it goes in the base executor class.

Maybe something like this in the base class
```
@contextmanager
def set_exec_refs(self):
    # init refs
    scm.set_ref(...)
    try:
        yield
    finally:
        # cleanup refs
        self.scm.remove_ref(...)
```

and then within each executor init_git you would have
```
def init_git():
    with self.init_exec_refs():
        # existing init_git code()
        ...
    # any additional executor specific cleanup handling here
    ...
```

Reply:
Only one exception in the `WorkspaceExecutor`, because it will remove these references after the task has been completed.